File: leetcode_py/p630.py
from typing import List
from collections import heapq


class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        courses.sort(key=lambda c: c[1])
        # 按照 结束时间进行排序

        q = list()

        total = 0
        for ti, di in courses:
            if total + ti <= di:
                # 这时候可以加进去
                heapq.heappush(q, -ti)  # 使用负数构建大根堆
            elif q and -q[0] > ti:  # 大根堆，最大的元素大于 ti
                # 这时候将大根堆，堆顶拿出来，因为有更优的方案
                total -= -q[0] - ti
                heapq.heappop(q)
                heapq.heappush(q, -ti)
        return len(q)

File: leetcode_C++/p630.cpp
#include <algorithm>
#include <queue>
#include <vector>
using namespace std;

class Solution {
public:
    int scheduleCourse(vector<vector<int>>& courses) {
        // 排序
        sort(courses.begin(), courses.end(), [](const auto& c0, const auto& c1) {
            return c0[1] < c1[1];
        });

        priority_queue<int> q;
        int total = 0;
        for (const auto& course : courses) {
            int ti = course[0], di = course[1];
            if (total + ti <= di) {
                q.push(ti);
                total += ti;
            } else if (!q.empty() && q.top() > ti) {
                total -= q.top() - ti;
                q.pop();
                q.push(ti);
            }
        }
        return q.size();
    }
};